### PR TITLE
Implement a common layer Aggr Executors

### PR DESCRIPTION
**Pull request by breeswish**
## What have you changed? (mandatory)

This PR introduces a common layer called `AggregationExecutor` for HashAggr, SimpleAggr and StreamAggr.

SimpleAggr is also re-organized to utilize `AggregationExecutor`.

- [x] Merge https://github.com/tikv/tikv/pull/4738 first
- [x] Merge https://github.com/tikv/tikv/pull/4731 first

## What are the type of the changes? (mandatory)

- Engineering (engineering change which doesn't change any feature or fix any issue)

## How has this PR been tested? (mandatory)

Existing unit tests.

### CODE REVIEW

File: src/coprocessor/dag/batch/executors/util/aggr_executor.rs [hunk 1]
```diff
@@ -0,0 +1,316 @@
...
+        let aggr_fn_len = aggr_defs.len();
+        assert!(aggr_fn_len > 0);
+
+        let src_schema = src.schema();
+        let src_schema_len = src_schema.len();
+
+        let mut schema = Vec::with_capacity(aggr_fn_len * 2);
+        let mut each_aggr_fn = Vec::with_capacity(aggr_fn_len);
+        let mut each_aggr_cardinality = Vec::with_capacity(aggr_fn_len);
+        let mut each_aggr_exprs = Vec::with_capacity(aggr_fn_len * 2);
```
Comment (sticnarf) [inline src/coprocessor/dag/batch/executors/util/aggr_executor.rs]: If we only support one parameter agg functions now, why we need a capacity of `aggr_fn_len * 2`?

File: src/coprocessor/dag/batch/executors/util/aggr_executor.rs [hunk 2]
```diff
@@ -0,0 +1,316 @@
...
+    /// Generates aggregation results.
+    ///
+    /// This function is ensured to be called at most once.
+    fn aggregate(&mut self) -> Result<LazyBatchColumnVec> {
+        let mut all_result_columns: Vec<_> = self
+            .entities
+            .all_result_column_types
+            .iter()
+            // Assume that there can be 1024 groups.
+            .map(|eval_type| VectorValue::with_capacity(1024, *eval_type))
```
[the remark this answers is not in the document]
Comment (breeswish) [reply src/coprocessor/dag/batch/executors/util/aggr_executor.rs]: It's simply an assumed value. We won't know this value before the aggregation is done, but anyway we need an estimation for the capacity.

File: src/coprocessor/dag/batch/executors/util/aggr_executor.rs (hunk 1 above)
Comment (breeswish) [reply src/coprocessor/dag/batch/executors/util/aggr_executor.rs]: nice catch. It's a mistake, should be only `aggr_fn_len`.

File: src/coprocessor/dag/batch/executors/util/aggr_executor.rs (hunk 2 above)
[the remark this answers is not in the document]
Comment (breeswish) [reply src/coprocessor/dag/batch/executors/util/aggr_executor.rs]: Update: actually we can know this value precisely. I will update this PR.

**Pull request merged**